Wangscape/tilegen/CartesianPower.h

```diff
@@ -0,0 +1,106 @@
+#pragma once
+#include "CartesianPowerIterator.h"
+#include <utility>
+
+namespace tilegen
+{
+
+template<typename InputIt>
+class CartesianPower
+{
+public:
+    static_assert(std::is_base_of<std::input_iterator_tag, 
+                                  typename std::iterator_traits<InputIt>::iterator_category>::value,
+                  "InputIt must be at least an InputIterator");
+    typedef CartesianPowerIterator<InputIt> const_iterator;
+
+    CartesianPower(InputIt first_, InputIt last_, size_t power_);
+    template<typename Container>
+    CartesianPower(const Container& container, size_t power_);
+
+    const_iterator cbegin() const;
+    const_iterator cend() const;
+    const_iterator begin() const;
+    const_iterator end() const;
+
+    size_t size() const;
+    std::pair<size_t, size_t> size_2d(size_t resolution) const;
+
+    InputIt first;
+    InputIt last;
```

Comment:
`first`, `base_size` (which replaces `last`), and `power` are all independent variables which can be changed without any requirements or unexpected side-effects. I don't see what problem would be fixed by putting them behind `get`/`set` methods.

(`CartesianPower::size()`/`size_2d()` are much simpler operations now that the `std::distance` call is removed, and they are only expected to be called once for each `CartesianPower`, so I don't think it would be worthwhile to cache those values).